File: Kotlin/src/main/kotlin/intermediate/boj_9370.kt
package intermediate

import java.util.PriorityQueue

private fun Dijk(grp: Array<ArrayList<Pair<Int, Int>>>, start: Int, end: Int): Long {
    val pq = PriorityQueue(Comparator<Pair<Int, Int>>{ a, b -> a.first - b.first })
    val distance = LongArray(grp.size) { Long.MAX_VALUE }

    pq.add(Pair(0, start))
    distance[start] = 0

    while (pq.isNotEmpty()) {
        val (dist, now) = pq.poll()

        if (distance[now] < dist) continue

        for ((next, weight) in grp[now]) {
            val cost = distance[now] + weight

            if (distance[next] > cost) {
                distance[next] = cost
                pq.add(Pair(cost.toInt(), next))
            }
        }
    }

    return distance[end]
}

fun boj_9370() = with(System.`in`.bufferedReader()) {
    repeat(readLine().toInt()) {
        val (n, m, t) = readLine()
            .split(" ")
            .map { it.toInt() }

        val grp = Array(n + 1) { ArrayList<Pair<Int, Int>>() }
        val lst = ArrayList<Int>()

        val (s, g, h) = readLine()
            .split(" ")
            .map { it.toInt() }

        repeat(m) {
            val (a, b, d) = readLine()
                .split(" ")
                .map { it.toInt() }

            grp[a].add(Pair(b, d))
            grp[b].add(Pair(a, d))
        }

        repeat(t) {
            lst.add(readLine().toInt())
        }

        // 출발지 -> 목적지의 거리 = 출발지 -> (교차로) -> 목적지의 거리가 같은 목적지
        val ans = ArrayList<Int>()

        for (end in lst) {
            val dist = Dijk(grp, s, end)
            val comp = Dijk(grp, s, g) + Dijk(grp, g, h) + Dijk(grp, h, end)
            val comp2 = Dijk(grp, s, h) + Dijk(grp, h, g) + Dijk(grp, g, end)
            if (dist == minOf(comp, comp2)) ans.add(end)
        }

        ans.sort()
        for (a in ans) print("$a ")
        println()
    }
}

File: Cpp/boj/Intermediate/boj_9370.cpp
//
// Created by seungsu on 12/2/22.
//

#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;
constexpr int INF = 1e8;

static int solve(vector<vector<pair<int, int>>>& grp, int start, int end) {
    priority_queue<pair<int, int>> pq;
    vector<int> distance(grp.size(), INF);

    pq.push(make_pair(0, start));
    distance[start] = 0;

    while (!pq.empty()) {
        pair<int, int> pr = pq.top();
        int now = pr.second;
        int dist = -pr.first;
        pq.pop();

        if (distance[now] < dist)
            continue;

        for (auto& tmp : grp[now]) {
            int next = tmp.first;
            int weight = tmp.second;
            int cost = distance[now] + weight;

            if (distance[next] > cost) {
                distance[next] = cost;
                pq.push(make_pair(-cost, next));
            }
        }
    }

    return distance[end];
}

void boj_9370() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        int n, m, t;    // 교차로, 도로, 목적지 후보 개수
        cin >> n >> m >> t;

        int s, g, h;    // 출발지, 교차로(g, h)
        cin >> s >> g >> h;

        vector<vector<pair<int, int>>> grp(n + 1);
        vector<int> lst(t);

        for (int i = 0; i < m; i++) {
            int a, b, d;
            cin >> a >> b >> d;

            grp[a].push_back(make_pair(b, d));
            grp[b].push_back(make_pair(a, d));
        }

        for (int i = 0; i < t; i++)
            cin >> lst[i];

        // start->end 거리가 min(start -> g -> h -> end, start -> h -> g -> end)과 같은 목적지들
        vector<int> ans;

        for (int k : lst) {
            int dist = solve(grp, s, k);
            int tmp1 = solve(grp, s, g) + solve(grp, g, h) + solve(grp, h, k);
            int tmp2 = solve(grp, s, h) + solve(grp, h, g) + solve(grp, g, k);

            if (dist == min(tmp1, tmp2))
                ans.push_back(k);
        }

        sort(ans.begin(), ans.end());
        for (int v : ans)
            cout << v << " ";

        cout << "\n";
    }
}